Expand BIP85 to include ECC key types

3rdIteration opened the pull request:
So I'm doing some work on extending BIP85 functionality on SeedSigner to include GPG key generation, file verification, encrypted messaging, etc, but the resource limitation on low power devices like the Raspberry Pi really makes RSA keys quite painful to use at any useful bit length. (RSA3072 takes 15-20 minutes for a primary key + 3 subs)

Basically I'm suggesting an expansion of the current functionality to add support for some common ECC curves and provide a bit of a process for future curves to be added in a way that doesn't break existing implementations. (The simplest solution is to just use ECC with key length on the existing RSA derivation, but that doesn't seem very consistent with how BIP85 works for other similar use cases where the derivation path should tell you everything you need to know about the output data/entropy/usage)

While it would have been nice to have a separate derivation level for key_type, that ship has probably sailed, so rather than add that, I am just suggesting a different applicationID for each key type. (Similar to how variations are handled with the existing password application type)

While I'm at it, I'm also suggesting some information about how additional and mixed key types should be handled.

Anyway, just keen to get some input on a way forward, can add some additional test vectors and stuff once others have provided some input about a preferred way forward.

3rdIteration committed:
Expand BIP85 to include ECC key types

jonatack commented on bip-0085.mediawiki:
```diff
@@ -360,36 +360,58 @@ OUTPUT
     - sub_key <code>2'</code> is usually used as SIGNATURE key
 
-Note on timestamps:
+Additional subkeys can be added to a primary key, including keys of a different key type, following the role pattern defined above, but the key_index MUST be incremented with each subkey.
```
Could this change break anything?

jonatack reviewed (commented):
Pinging @akarve and @scgbckbone for feedback, BIP85 is in use and in Final status (https://github.com/bitcoin/bips/pull/1676), so no potentially breaking changes may be introduced.

akarve commented on bip-0085.mediawiki:
```diff
@@ -360,36 +360,58 @@ OUTPUT
+===GPG Keys===
 
-Application number: 828365'
+Application number is dependant on the key type.
```
If these are all GPG keys why not just add a segment for the key type? That would be more consistent with the rest of 85 so:
```
{one app number for all}/{key_type}/{bits}/{index}
```

akarve commented on bip-0085.mediawiki:
```diff
@@ -360,36 +360,58 @@ OUTPUT
+===GPG Keys===
 
-Application number: 828365'
+Application number is dependant on the key type.
```
I would recommend a single application number and then extend the path or different key types. That's more consistent with the rest of BIP85:
```
{same_app_number_for_all}/{key_type}/{key_bits}/{key_index}
```

3rdIteration replied:
This would be the neatest solution

akarve commented on bip-0085.mediawiki:
```diff
@@ -360,36 +360,58 @@ OUTPUT
 
- - Main key <code>m/83696968'/828365'/{key_bits}'/{key_index}'</code>
- - Sub keys:  <code>m/83696968'/828365'/{key_bits}'/{key_index}'/{sub_key}'</code>
+ - Primary key <code>m/83696968'/{key_type}'/{key_bits}'/{key_index}'</code>
```
Here and elsewhere these list nested code blocks don't render as expected (... > View file)

akarve commented on bip-0085.mediawiki:
```diff
@@ -360,36 +360,58 @@ OUTPUT
     - sub_key <code>2'</code> is usually used as SIGNATURE key
 
-Note on timestamps:
+Additional subkeys can be added to a primary key, including keys of a different key type, following the role pattern defined above, but the key_index MUST be incremented with each subkey.
```
As for mixing types I would discourage that unless there's a really good reason. For one thing are we nesting paths at that point? A worked example of mixed subkeys would help.

akarve replied:
> the use-case for this is basically scenarios where someone may want a strong and long-lived primary key that for their identity, but use different keys for the actual signing and stuff. (Whether due to limitations on smartcards, etc) For example, someone may have an ED25119 primary key and have subkeys use something like P256 or RSA subkeys for compatibility reasons.

reasonable enough. at that point why not turtles all the way down, e.g. the following or something like it?

```
{same_app_number_for_all}/
{key_type}/{key_bits}/{key_index}/
{key_type}/{key_bits}/{sub_key_index}/
...
```

akarve commented on bip-0085.mediawiki:
```diff
@@ -360,36 +360,58 @@ OUTPUT
 * DERIVED PWD=_s`{TW89)i4`
 
-===RSA===
+===GPG Keys===
```
Definitely need test vectors and expected outputs for all applications and sub applications. I can also add it to the reference implementation but won't be able to get to that right away. PRs welcome. https://github.com/akarve/bipsea

akarve commented on bip-0085.mediawiki:
```diff
@@ -360,36 +360,58 @@ OUTPUT
 
 However, depending on available slots on the smart-card, and preferred policy, the CERTIFY capable key MAY be flagged with CERTIFY and SIGNATURE capabilities and loaded into the SIGNATURE capable slot (for example where the smart-card has only three slots and the CERTIFY capability is required on the same card). In this case, the SIGNATURE capable sub-key would be disregarded because the CERTIFY  ...
 
+
```
```suggestion
```

3rdIteration committed:
Update bip-0085.mediawiki

Co-authored-by: Jon Atack <jon@atack.com>

3rdIteration committed:
Update bip-0085.mediawiki

Co-authored-by: Aneesh Karve <akarve@users.noreply.github.com>

3rdIteration commented:
Are there any that aren't addressed? (In terms of the other PR, I agree that option C makes the most sense)

It also gives more scope to make some changes that make this neater overall, even if it deviates from the original RSA-Only implementation.

jonatack commented:
> Are there any that aren't addressed? (In terms of the other PR, I agree that option C makes the most sense)

@3rdIteration the changelog update, not sure for the rest of @akarve's feedback

jonatack commented:
#1967 has been merged, so this can now be rebased to master and the changelog updated.

3rdIteration committed:
Update bip-0085.mediawiki

3rdIteration committed:
Update bip-0085.mediawiki

3rdIteration commented on bip-0085.mediawiki:
```diff
@@ -360,31 +360,57 @@ OUTPUT
+|-
+| ECC(Curve25519)
+| 1'
+|-
```
@copilot code review[agent] Can you add a note for the ECC key_types that mentions which key_bits lengths are valid

3rdIteration committed:
incorporate improment suggestions

3rdIteration committed:
Fix typos

3rdIteration committed:
Enhance ECC and RSA key generation guidelines

Expanded the section on ECC scalar derivation and RSA key generation, detailing the use of BIP85-DRNG and compliance requirements.

3rdIteration committed:
Add test vectors for key types in bip-0085

Added test vectors for various key types including RSA, Curve25519, secp256k1, NIST, and Brainpool.

3rdIteration committed:
Fix heading format for Test vectors section

3rdIteration committed:
Merge branch 'master' into BIP85-ExpandCurveTypes

3rdIteration committed:
Apply suggestion from @akarve

Co-authored-by: Aneesh Karve <akarve@users.noreply.github.com>

akarve commented on bip-0085.mediawiki:
```diff
@@ -388,6 +388,179 @@ GPG capable smart-cards SHOULD be loaded as follows: The encryption slot SHOULD
+
+For key_types with a fixed curve (1 and 2), key_bits MUST be 256. Any other value is invalid.
+
+BIP85-DRNG MUST be used as the input RNG function when key generation requires more than 64 bytes of random input. This applies to NIST P-521 (key_type=3, key_bits=521). All other supported ECC key types use the 64-byte HMAC output directly.
```
Please break all lines at 80-100 characters or less for easier diffs and terminal reads.

akarve commented on bip-0085.mediawiki:
```diff
@@ -388,6 +388,179 @@ GPG capable smart-cards SHOULD be loaded as follows: The encryption slot SHOULD
+
+<b>ECC Scalar Derivation</b>
+
+For curves with base length ≤ 64 bytes (all except NIST P-521), the private scalar is derived from the first <code>baselen</code> bytes of the 64-byte HMAC-SHA512 entropy output, interpreted as a big-endian unsigned integer. If the resulting scalar is zero or greater than the curve order, apply the fallback: <code>scalar = (scalar % (order - 1)) + 1</code>.
```
In order to avoid sampling bias we need to do the "same for all" idea not just for the fallback but with no special cases anywhere here or below. And cite this as FIPS 186-5 A.2.1. https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-5.pdf

Without this the residue classes do not distribute evenly.

https://github.com/bitcoin/bips/pull/1968#discussion_r3036037609

akarve commented:
This is almost ready. See updated comments, esp. on where we need to apply FIPS 186-5 to avoid bias.

vitrixLab commented:
**Promising, not merge-ready.**

`key_type` / `key_bits` do not yet unambiguously identify the ECC curve and OpenPGP algorithm/role for all supported families, especially NIST/Brainpool and Curve25519.

Required for deterministic interoperability:

1. Normative curve/algorithm/role mappings
2. Explicit valid-value constraints per derivation component
3. Complete reproducible test vectors for every supported application/key type

Primary gate: **specification determinism / interoperability**
